### PR TITLE
Disable auto-setting server as favorite on connect; display "Add Favorite" button

### PR DESCRIPTION
**Pull request by ClobberXD**
Currently, a server is automatically added to one's favourites when they connect to it (or even attempt to connect to it). I think this isn't a good idea, as players would be trying out new servers, and might not ever visit them again, and yet, those servers will unnecessarily be added to their favourites.

This PR disables "auto-favouriting" all servers the player joins, and provides an "Add Favorite" button. Depending on whether a server is marked as favourite, there will be a "Del. Favorite" button or an "Add Favorite" button. A server will be marked as favorite only when a player explicitly adds it by pressing the "Add Favorite" button.

### CODE REVIEW

**Comment by ClobberXD** [comment]
Updated first post with the remaining issues. PR will remain WIP until these are taken care of, but I could always use some help in fixing these. :)

**Comment by paramat** [comment]
PR concept makes good sense, current behaviour doesn't.

**Comment by Fixer-007** [comment]
I was frustrated with this for a long time, but was too lazy to create an issue, thank you for doing this PR, +1

**Comment by ClobberXD** [comment]
I think I got the hang of this now - seems to work. fine now. Ready for testing and review. :)

**Comment by pgimeno** [comment]
I sort of like the current behaviour. Could this be made optional? At least until #8318 is implemented.

**Comment by jastevenson303** [comment]
The current is more like "recently visited" and less like "favorited."
Ideally, I think, both would be present.

**Comment by ClobberXD** [comment]
Instead of making this optional for the time-being, putting this PR on hold seems to be an easier approach to me :)

File: doc/menu_lua_api.txt [hunk 1]
```diff
@@ -17,12 +17,12 @@ Gamedata
...
-	address        = <IP/adress>,
-	port           = <port>,
-	selected_world = <index>, -- 0 for client mode
-	singleplayer   = <true/false>,
+    playername     = <name>,
+    password       = <password>,
+    address        = <IP/adress>,
+    port           = <port>,
+    selected_world = <index>, -- 0 for client mode
+    singleplayer   = <true/false>,
```
Comment (pgimeno) [inline doc/menu_lua_api.txt]: Is this a pure text file, or a .md file renamed to .txt in the style of lua_api.txt? If the latter, then the formatting is wrong all over the place, and this change is incomplete. See https://notabug.org/pgimeno/Gists/src/minetest-clobberxd---menu_lua_api/menu_lua_api.md

I suggest to either leave the formatting alone, or add a separate cleanup commit or even a new PR to clean it up.

**Comment by DS-Minetest** [comment]
You might want to have a server in the serverlist (=favorite list) that is not in the public server list (=online favorites). This feature is important and was there before we got the online favorites. It should not be simply removed.

I think, the whole change should be done in one PR.
(Rename favorites to visited and add real favorites or something.)

**Comment by pgimeno** [comment]
I'm working on a patch based on this one.

File: doc/menu_lua_api.txt [hunk 2]
```diff
@@ -98,74 +98,86 @@ core.file_open_dialog(formname,caption)
...
 }
-core.delete_favorite(id, location) -> success
+
+- core.add_favorite(server)
+  - Adds a server to the local favorites list
+  - `server`: {
+        `name` (server name)
+        `address`
+        `port`
+        `description`
```
Comment (DS-Minetest) [inline doc/menu_lua_api.txt]: This is no real table and will cause confusion.
One might think that it's a list (eg. `{"mycoolserver", "bla.net",30000,"come and join"}` but the commas are missing.

It should be something like:
```
{
  name    = <server name>,
  address = <address of the server>,
  ...
}
```

**Comment by pgimeno** [comment]
@ClobberXD Mind if I take your code but removing most changes to menu_lua_api.txt, as a basis for a new patch?

**Comment by DS-Minetest** [comment]
There are many changes that don't have too much to do with favorites not being marked on join (eg. trying to delete a server from the public serverlist (which does nothing) is here removed). I think, these changes should be made in separate commits.

**Comment by pgimeno** [comment]
@DS-Minetest I believe the list is only of favourite servers, therefore removing an element from the list effectively removes the favourite.

File: doc/menu_lua_api.txt [hunk 3]
```diff
@@ -98,74 +98,86 @@ core.file_open_dialog(formname,caption)
...
+    }
+
+- core.delete_favorite(idx)
+  - Removes a server from the local favorites list. Returns `true` on success.
```
Comment (DS-Minetest) [inline doc/menu_lua_api.txt]: I think, the ``Returns `true` on success.`` is wrong.

**Comment by DS-Minetest** [comment]
>eg. trying to delete a server from the public serverlist (which does nothing) is here removed

> @DS-Minetest I believe the list is only of favourite servers, therefore removing an element from the list effectively removes the favourite.

Yes, but removing an element from the online favorite list (= public server list) doesn't really work. [I've tested it.](https://github.com/DS-Minetest/minetest/tree/mainmenu_online_bla)
This PR changes `delete_favorite` to only delete servers from the local list. This is good but should be done in a separate commit.
I'm sorry for expressing myself so unclear. >_<

File: builtin/mainmenu/tab_online.lua [hunk 4]
```diff
@@ -229,6 +233,17 @@ local function main_button_handler(tabview, fields, name, tabdata)
 		return true
 	end
 
+	if fields.btn_add_favorite then
+		core.add_favorite({
+			name        = gamedata.servername,
+			address     = gamedata.address,
+			port        = gamedata.port,
+			description = gamedata.description
```
Comment (pgimeno) [inline builtin/mainmenu/tab_online.lua]: This is causing a crash. `gamedata` doesn't seem to hold this information. It is in `serverlist[tabdata.fav_selected]` (if `tabdata.fav_selected` is not nil or 0).

File: src/serverlist.cpp [hunk 5]
```diff
@@ -115,9 +115,7 @@ bool deleteEntry(const ServerListSpec &server)
 	std::string path = ServerList::getFilePath();
 	std::ostringstream ss(std::ios_base::binary);
 	ss << ServerList::serialize(serverlist);
-	if (!fs::safeWriteToFile(path, ss.str()))
-		return false;
-	return true;
+	return !fs::safeWriteToFile(path, ss.str()) ? true : false;
```
Comment (pgimeno) [inline src/serverlist.cpp]: This is the wrong way around. I suggest to revert this change.

File: src/serverlist.cpp [hunk 6]
```diff
@@ -134,10 +132,7 @@ bool insert(const ServerListSpec &server)
 	std::string path = ServerList::getFilePath();
 	std::ostringstream ss(std::ios_base::binary);
 	ss << ServerList::serialize(serverlist);
-	if (!fs::safeWriteToFile(path, ss.str()))
-		return false;
-
-	return true;
+	return !fs::safeWriteToFile(path, ss.str()) ? true : false;
```
Comment (pgimeno) [inline src/serverlist.cpp]: This is the wrong way around. I suggest to revert this change. There's no point in touching this file just for rewriting the style of the condition.

File: builtin/mainmenu/tab_online.lua (hunk 4 above)
Comment (DS-Minetest) [reply builtin/mainmenu/tab_online.lua]: There's a `gamedata.serverdescription`.

File: src/serverlist.cpp (hunk 5 above)
Comment (DS-Minetest) [reply src/serverlist.cpp]: Can't this simply be `return fs::safeWriteToFile(path, ss.str())`?

File: builtin/mainmenu/tab_online.lua (hunk 4 above)
Comment (pgimeno) [reply builtin/mainmenu/tab_online.lua]: It's empty at that point. Also, I mean all four fields, but I didn't want to comment on all four lines :)

Have you tried the patch? It crashes for me due to this data being empty.

File: src/serverlist.cpp (hunk 5 above)
Comment (pgimeno) [reply src/serverlist.cpp]: Yes, but not point in changing the file just for that, which is why I suggested a reversal.

File: src/serverlist.cpp (hunk 6 above)
Comment (ClobberXD) [reply src/serverlist.cpp]: Will revert both these changes.

**Comment by ClobberXD** [comment]
@pgimeno Feel free to use this PR as a base for another patch - I have no issues with it. :)

**Comment by ClobberXD** [comment]
Is someone working on #8318? If no one's working on that PR, what'd happen to this PR?

**Comment by ClobberXD** [comment]
\*bump\*

Core dev inputs needed. This PR was put on hold because pgimeno was planning to implement #8318 first, but has now stopped working on it (see prev. post). What's the fate of this PR now?

This PR is still not perfect. But this PR is still a significant step in the right direction, even without #8318.

**Comment by paramat** [comment]
I closed #8318 as it is abandoned. So just continue as if it does not exist.

**Comment by ClobberXD** [comment]
Thanks, paramat. Will continue with this.

**Comment by paramat** [comment]
To clarify, i think this change is still desirable even without the 'recently visited' ordering that #8318 implemented.

**Comment by Fixer-007** [comment]
Any news on this?

**Comment by ClobberXD** [comment]
I'd say this might take a week or two to attend to. But fortunately for me, there's only a small amount of cleaning up to do, and the PR would be fresh as a daisy - there aren't any mammoth merge conflicts I have to deal with. :)

**Comment by sfan5** [comment]
If this is added, how would I add a server this is not on the server list to my favorites?

**Comment by ClobberXD** [comment]
> If this is added, how would I add a server this is not on the server list to my favorites?

This PR removes the ability to do this; this is a very valid concern.

I'm thinking of re-adding the auto-favourite-upon-connect functionality, but only if a server isn't on the server list. Thoughts?

**Comment by sfan5** [comment]
> I'm thinking of re-adding the auto-favourite-upon-connect functionality, but only if a server isn't on the server list. Thoughts?

Sounds fine to me.

**Comment by SmallJoker** [comment]
Non-listed servers should be added to the favourites automatically. So yes, that suggestion sounds good.

> I sort of like the current behaviour. 

> though I still prefer the overall "favorite-on-connect" feature to stay as is.

Me as well. It's helpful to keep track of the servers I usually play on. Thus I'm not really in favour of this PR. A setting could be added for sure, but should then also be integrated directly into the mainmenu (checkbox, perhaps).

**Comment by ClobberXD** [comment]
Then the *correct* solution would be to have two categories "visited" and "favourites". While there's support for this PR, I do also see some opposition - I'd prefer to not merge this PR until we arrive upon  an 'ideal' solution.
